### PR TITLE
Diagnostics flush parameters to not open/close each dump

### PR DESCRIPTION
**Pull request by PhilipDeegan**

### CODE REVIEW

File: tests/diagnostic/test_diagnostics.ipp [hunk 1]
```diff
@@ -81,6 +81,7 @@ void allFromPython_test(Simulator&& sim, std::string out_dir)
     using Hierarchy   = typename Simulator::Hierarchy;
 
     sim.dump(*sim.dMan);
+    sim.dMan.reset(); // flush h5files
```
Comment (PhilipDeegan) [inline tests/diagnostic/test_diagnostics.ipp:84]: this is maybe not ideal

File: src/diagnostic/detail/h5typewriter.h [hunk 2]
```diff
@@ -41,7 +42,15 @@ class H5TypeWriter : public PHARE::diagnostic::TypeWriter
         std::size_t maxLevel)
         = 0;
 
-    virtual void finalize(DiagnosticProperties& diagnostic) = 0;
+    void finalize(DiagnosticProperties& diagnostic)
+    {
+        ++dumpIdx_;
+        if (diagnostic.flushEvery > 0 and diagnostic.flushEvery % dumpIdx_ == 0)
```
Comment (nicolasaunai) [inline src/diagnostic/detail/h5typewriter.h:48]: I think the first condition is checked already in python thus redundant here?

File: pyphare/pyphare/pharein/__init__.py [hunk 3]
```diff
@@ -195,6 +196,7 @@ def as_paths(rb):
         name_path = type_path + diag.name
         add(name_path + "/" + 'type' , diag.type)
         add(name_path + "/" + 'quantity' , diag.quantity)
+        add_size_t(name_path + "/" + "flush_every", diag.flush_every)
```
Comment (nicolasaunai) [inline pyphare/pyphare/pharein/__init__.py:199]: we have to be cautious that "flush every" could be interpreted as "every N time step" although it really means (if I understand correctly) "flush every N dump"

File: tests/diagnostic/test_diagnostics.h [hunk 4]
```diff
@@ -207,10 +207,11 @@ struct Hi5Diagnostic
...
         PHARE::initializer::PHAREDict dict;
-        dict["name"]      = quantity;
-        dict["type"]      = type;
-        dict["quantity"]  = quantity;
-        dict["time_step"] = double{1};
+        dict["name"]        = quantity;
+        dict["type"]        = type;
+        dict["quantity"]    = quantity;
+        dict["time_step"]   = double{1};
+        dict["flush_every"] = std::size_t{0};
```
Comment (nicolasaunai) [inline tests/diagnostic/test_diagnostics.h:214]: 0? what does it mean?

File: tests/diagnostic/test_diagnostics.ipp (hunk 1 above)
Comment (nicolasaunai) [reply tests/diagnostic/test_diagnostics.ipp:84]: can you elaborate a bit please?

File: tests/simulator/test_diagnostic_timestamps.py [hunk 5]
```diff
@@ -107,9 +107,10 @@ def test_dump_diags_timestamps(self):
                 quantity=quantity,
                 write_timestamps=timestamps,
                 compute_timestamps=timestamps,
+                flush_every=0,
```
Comment (nicolasaunai) [inline tests/simulator/test_diagnostic_timestamps.py:110]: same, 0 ?

**Review by nicolasaunai — APPROVED** [review]
merge when you see fit

**Pull request merged**